engine/table/src/main/java/io/deephaven/engine/table/impl/SelectOrUpdateListener.java

```diff
@@ -0,0 +1,138 @@
+package io.deephaven.engine.table.impl;
+
+import io.deephaven.engine.rowset.TrackingRowSet;
+import io.deephaven.engine.rowset.WritableRowSet;
+import io.deephaven.engine.table.ModifiedColumnSet;
+import io.deephaven.engine.table.TableUpdate;
+import io.deephaven.engine.table.impl.perf.UpdatePerformanceTracker;
+import io.deephaven.engine.table.impl.select.analyzers.SelectAndViewAnalyzer;
+import io.deephaven.engine.table.impl.util.AsyncClientErrorNotifier;
+import io.deephaven.engine.updategraph.LogicalClock;
+import io.deephaven.engine.updategraph.TerminalNotification;
+import io.deephaven.engine.updategraph.UpdateGraphProcessor;
+import io.deephaven.engine.util.systemicmarking.SystemicObjectTracker;
+
+import java.io.IOException;
+import java.util.BitSet;
+import java.util.Map;
+
+/**
+ * A Shift-Aware listener for Select or Update. It uses the SelectAndViewAnalyzer to calculate how columns affect other
+ * columns, then creates a column set transformer which will be used by onUpdate to transform updates.
+ */
+class SelectOrUpdateListener extends BaseTable.ListenerImpl {
+    private final QueryTable dependent;
+    private final TrackingRowSet resultIndex;
+    private final ModifiedColumnSet.Transformer transformer;
+    private final SelectAndViewAnalyzer analyzer;
+
+    private long initialNotificationStep = NotificationStepReceiver.NULL_NOTIFICATION_STEP;
+    private long finalNotificationStep = NotificationStepReceiver.NULL_NOTIFICATION_STEP;
+    final BitSet completedColumns = new BitSet();
+    final BitSet allNewColumns = new BitSet();
+    final boolean enableParallelUpdate;
+
+    /**
+     * @param description Description of this listener
+     * @param parent The parent table
+     * @param dependent The dependent table
+     * @param effects A map from a column name to the column names that it affects
+     */
+    SelectOrUpdateListener(String description, QueryTable parent, QueryTable dependent, Map<String, String[]> effects,
+            SelectAndViewAnalyzer analyzer) {
+        super(description, parent, dependent);
+        this.dependent = dependent;
+        this.resultIndex = dependent.getRowSet();
+
+        // Now calculate the other dependencies and invert
+        final String[] parentNames = new String[effects.size()];
+        final ModifiedColumnSet[] mcss = new ModifiedColumnSet[effects.size()];
+        int nextIndex = 0;
+        for (Map.Entry<String, String[]> entry : effects.entrySet()) {
+            parentNames[nextIndex] = entry.getKey();
+            mcss[nextIndex] = dependent.newModifiedColumnSet(entry.getValue());
+            ++nextIndex;
+        }
+        transformer = parent.newModifiedColumnSetTransformer(parentNames, mcss);
+        this.analyzer = analyzer;
+        this.enableParallelUpdate =
```

Comment:
1. Missing force
2. Why not just pass this in, since the calling code already knows the right answer? Oh, you answered that. Fine. Just fine.